eclipse/plugin/src/nl/tudelft/watchdog/eclipse/logic/ui/listeners/EclipseMarkupModelListener.java

```diff
@@ -0,0 +1,213 @@
+		 * the difference between them. Delegate this implementation to
+		 * {@link MarkerBackTrackingAlgorithm} which is an abstraction around
+		 * this algorithm.
+		 */
+		private boolean visit(IResource resource, boolean shouldComputeDiff) throws CoreException {
+			if (!resource.exists()) {
+				return false;
+			}
+
+			IPath filePath = resource.getFullPath();
+			List<MarkerHolder> oldMarkers = oldFileMarkers.get(filePath);
+			List<MarkerHolder> currentMarkers =
+					Arrays.stream(resource.findMarkers(IMarker.PROBLEM, true, IResource.DEPTH_ZERO))
+						  .map(MarkerHolder::fromIMarker)
+						  .sorted()
+						  .collect(Collectors.toList());
+
+			if (shouldComputeDiff && oldMarkers == null) {
+				oldMarkers = Collections.emptyList();
+			}
+
+			if (oldMarkers != null) {
+				new MarkerBackTrackingAlgorithm(oldMarkers, currentMarkers)
+						.computeMemoizationTable()
+						.traverseMemoizationTable();
+			}
+
+			currentFileMarkers.put(filePath, currentMarkers);
+
+			return true;
+		}
+	}
+
+	/**
+	 * Because an {@link IMarker} throws exceptions on {@link IMarker#getAttribute(String)} if it is removed,
+	 * we have to extract all useful information at the first moment we encounter it, instead of invoking
+	 * these methods once the marker is removed. At this moment, we are only interested in its message,
+	 * as all other attributes are non-unique (yes even the ID is not unique). The only way to keep track
+	 * of messages in a somewhat unique manner is its message, which does include private information.
+	 * Make sure that when you use this message, the data is anonymized.
+	 */
+	static class MarkerHolder implements Comparable<MarkerHolder> {
+		private String message;
+		private int lineNumber;
+
+		static MarkerHolder fromIMarker(IMarker marker) {
+			MarkerHolder holder = new MarkerHolder();
+			holder.message = marker.getAttribute(IMarker.MESSAGE, "");
+			holder.lineNumber = marker.getAttribute(IMarker.LINE_NUMBER, 0);
+			return holder;
+		}
+
+		@Override
+		public boolean equals(Object other) {
+			if (other instanceof MarkerHolder) {
+				return this.message.equals(((MarkerHolder) other).message);
+			}
+			return false;
+		}
+
+		@Override
+		public int compareTo(MarkerHolder other) {
+			return Integer.compare(this.lineNumber, other.lineNumber);
+		}
+	}
+
+	/**
+	 * An implementation of an algorithm to solve the longest common subsequence problem.
+	 * The implementation relies on a memoization table to compute (in O(n * m)) all lengths
+	 * of cost of modifications to the lists. It then traverses through this matrix
+	 * and based on the cost either triggers a deletion or an addition.
+	 *
+	 * There are multiple viable implementations of this diffing
+	 * algorithm and this abstraction allows us to later change the actual
+	 * implementation without breaking the usage contract.
+	 *
+	 * @see https://en.wikipedia.org/wiki/Longest_common_subsequence_problem
+	 */
+	class MarkerBackTrackingAlgorithm {
```

Comment:
visibility?

Reply:
This class is intended to be an implementation detail of the listener and thus should not be `public`. Could be `private`, but that breaks the Javadoc `{@link}` in Eclipse :cry: